File: backend/schemas/pricing.py
"""
backend/schemas/pricing.py
--------------------------
Pydantic models for Pricing Intelligence analytics and plan exploration.
"""

from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


class PricingModelStats(BaseModel):
    model: str
    count: int
    percentage: float


class PriceAmountDistribution(BaseModel):
    min_price: float | None = None
    max_price: float | None = None
    avg_price: float | None = None
    median_price: float | None = None
    p25_price: float | None = None
    p75_price: float | None = None


class IntervalBreakdown(BaseModel):
    billing_interval: str
    plan_count: int
    percentage: float


class FreeTrialStats(BaseModel):
    has_trial_count: int
    no_trial_count: int
    trial_percentage: float
    popular_durations: dict[int, int] = Field(default_factory=dict)


class PricingOverviewResponse(BaseModel):
    total_apps: int
    total_pricing_plans: int
    model_breakdown: list[PricingModelStats]
    price_distribution: PriceAmountDistribution
    interval_breakdown: list[IntervalBreakdown]
    free_trial_stats: FreeTrialStats


class PlanExplorerItem(BaseModel):
    id: int
    app_id: int
    app_slug: str
    app_name: str
    plan_name: str
    price_amount: float | None = None
    currency: str = "USD"
    billing_interval: str = "monthly"
    free_trial_days: int | None = None
    features: str | list[str] | None = None
    created_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/pricing.ts
/**
 * frontend/src/api/pricing.ts
 * API calls for Pricing Intelligence and Plan Explorer.
 */

import { apiClient } from './client';
import type { PlanExplorerItem } from '../types';
import type { PaginatedResponse } from './apps';

export interface PricingModelStats {
  model: string;
  count: number;
  percentage: number;
}

export interface PriceAmountDistribution {
  min_price: number | null;
  max_price: number | null;
  avg_price: number | null;
  median_price: number | null;
  p25_price: number | null;
  p75_price: number | null;
}

export interface IntervalBreakdown {
  billing_interval: string;
  plan_count: number;
  percentage: number;
}

export interface FreeTrialStats {
  has_trial_count: number;
  no_trial_count: number;
  trial_percentage: number;
  popular_durations: Record<string, number>;
}

export interface PricingOverviewResponse {
  total_apps: number;
  total_pricing_plans: number;
  model_breakdown: PricingModelStats[];
  price_distribution: PriceAmountDistribution;
  interval_breakdown: IntervalBreakdown[];
  free_trial_stats: FreeTrialStats;
}

export interface PricingPlanQueryParams {
  q?: string;
  app_slug?: string;
  billing_interval?: string;
  min_price?: number;
  max_price?: number;
  has_free_trial?: boolean;
  sort_by?: 'price' | 'plan_name' | 'app' | 'newest';
  sort_order?: 'asc' | 'desc';
  page?: number;
  limit?: number;
}

export async function fetchPricingOverview(): Promise<PricingOverviewResponse> {
  return apiClient<PricingOverviewResponse>('/pricing/overview');
}

export async function fetchPricingPlans(
  params?: PricingPlanQueryParams
): Promise<PaginatedResponse<PlanExplorerItem>> {
  return apiClient<PaginatedResponse<PlanExplorerItem>>('/pricing/plans', undefined, params as any);
}
